sdks/java/extensions/timeseries/src/main/java/org/apache/beam/sdk/extensions/timeseries/joins/BiTemporalStreams.java

```diff
@@ -0,0 +1,464 @@
+            }
+          }
+          // Increase the processed Trade Counter to check for GC LIMIT
+          ++processedLeftStreamCount;
+        }
+
+        // Check if there is more work left in the leftStreamValue queue
+        // Create timer to fire if there is
+        // If the number of old left stream values in the queue is @ LEFT_STREAM_GC_LIMIT then do GC
+        // If no work left delete the Trade Queue
+
+        if (leftStreamValue != null
+            && leftStreamValue.getTimestamp().getMillis() > processTimerTimestamp) {
+          processTimer.set(leftStreamValue.getTimestamp());
+          timerTimestamp.write(leftStreamValue.getTimestamp().getMillis());
+
+          // Clear down queue if we are at LEFT_STREAM_GC_LIMIT
+          if (processedLeftStreamCount >= LEFT_STREAM_GC_LIMIT) {
+            List<TimestampedValue<V1>> leftStreamValues =
+                leftStreamValueList.subList(
+                    processedLeftStreamCount - 1, leftStreamValueList.size());
+            leftStream.clear();
+            leftStreamValues.forEach(i -> leftStream.add(i));
+            putSortedList(otc.window(), leftStreamValues, leftCache, dataKey.read());
+            LOG.debug(
+                "GC - Partial WM is {} size of queue is {} ",
+                onTimerTimestamp,
+                leftStreamValues.size());
+          }
+        } else {
+          timerTimestamp.clear();
+          putSortedList(otc.window(), null, leftCache, dataKey.read());
+          putSortedList(otc.window(), null, rightCache, dataKey.read());
+          LOG.info("GC - Full WM is {}", onTimerTimestamp);
+        }
+
+        lastProcessedLeftStreamTimestampState.write(lastProcessedLeftStreamTimestamp);
+      }
+
+      // Get the sorted list from our Cache
+      public <V> List<TimestampedValue<V>> getSortedList(
+          BoundedWindow window,
+          BagState<TimestampedValue<V>> stream,
+          Map<WindowedValue, List<TimestampedValue<V>>> cache,
+          K elementKey) {
+
+        // Store this window-key combination in the Cache
+        WindowedValue cacheKey = getCacheKeyForWindow(elementKey, window.maxTimestamp(), window);
+
+        List<TimestampedValue<V>> sortedBundleList =
+            cache.computeIfAbsent(cacheKey, key -> createAndLoadSortedList(window, stream));
+
+        cache.put(cacheKey, sortedBundleList);
+
+        return sortedBundleList;
+      }
+
+      // Build the sorted list if its not in the Cache.
+      public <V> List<TimestampedValue<V>> createAndLoadSortedList(
+          BoundedWindow window, BagState<TimestampedValue<V>> stream) {
+
+        List<TimestampedValue<V>> sortedBundleList = new ArrayList<>();
+        final List<TimestampedValue<V>> data = sortedBundleList;
+        stream.read().forEach(t -> data.add(t));
+        data.sort(Comparator.comparing(t -> t.getTimestamp()));
+        sortedBundleList = data;
+
+        LOG.debug("Catch miss for maxTime {} ", window.maxTimestamp());
```

Comment:
```suggestion
        LOG.debug("Cache miss for maxTime {} ", window.maxTimestamp());
```

Reply:
The potential for OOM is not just in batch it can also happen in stream, dependent on the data size. For the datasets that we first explored this use case with however largemem workers was able to deal with the data volumes even with very large windows. 

One solution to reduce the chance of a OOM, would be to wrap an outer sharding layer to this transform. This would shard up the right hand data into smaller time segments. And the left hand side would be sent to all-possible time shards on the right hand side. With multiple matches evaluated in downstream transform. 

All of that could also be made transparent to the user of course, but I had not yet gone down that route in this @experimental class. It would be good to see if folks find it useful and to get back data on expected data sizes before adding in the complexity. Also I understand folks are looking into a potential sorted map state, which would allow for the whole of this cache code to be deleted :-)

Reply:
It is true, that it can OOM on stream, but the chances are pretty m uch lower, than on stream. Essentially, watermark would have to stop being forwarded for long enough. But - even if this happens, checkpointing and recovery mechanisms of streaming engines would eventually make the watermark move and the code would have a chance to complete. Once this fails on batch, it will fail forever. And there is practically no reason for that, because on batch the input can be sorted, and watermarks can be updated gradually (instead of having two states - min and max). That would be way more efficient.